climpred/bootstrap.py

```diff
@@ -173,81 +212,131 @@ def bootstrap_perfect_model(ds,
+    pers = xr.concat(pers, dim='bootstrap')
+
+    init_ci = _distribution_to_ci(init, ci_low, ci_high)
+    uninit_ci = _distribution_to_ci(uninit, ci_low, ci_high)
+    pers_ci = _distribution_to_ci(pers, ci_low_pers, ci_high_pers)
+
+    p_uninit_over_init = _pvalue_from_distributions(uninit, init)
+    p_pers_over_init = _pvalue_from_distributions(pers, init)
 
     # calc skill
     init_skill = compute_perfect_model(
-        ds, control, metric=metric, comparison=comparison, running=running,
-        reference_period=reference_period)
+        ds, metric=metric, comparison=comparison)
+    uninit_skill = uninit.mean('bootstrap')
+    pers_skill = compute_persistence(
+        ds, control, metric=metric)
+
+    # somehow there may be a member dim, which lets concat crash
+    if 'member' in init_skill:
+        del init_skill['member']
+    # wrap results together in one dataarray
+    skill = xr.concat([init_skill.squeeze(), uninit_skill, pers_skill], 'i')
+    skill['i'] = ['init', 'uninit', 'pers']
+
+    # probability that i beats init
+    p = xr.concat([p_uninit_over_init, p_pers_over_init], 'i')
+    p['i'] = ['uninit', 'pers']
+
+    # ci for each skill
+    ci = xr.concat([init_ci, uninit_ci, pers_ci],
+                   'i').rename({'quantile': 'results'})
+    ci['i'] = ['init', 'uninit', 'pers']
+
+    results = xr.concat([skill, p], 'results')
+    results['results'] = ['skill', 'p']
+    # (RXB) Drop any spurious coordinates that came along with results due to
+    # input ds. You can't concatenate if results and ci don't exactly match
+    # in coordinates. Maybe we can create a decorator to check this before
+    # anytime we merge.
+    if set(results.coords) != set(ci.coords):
+        res_drop = [c for c in results.coords if c not in ci.coords]
+        ci_drop = [c for c in ci.coords if c not in results.coords]
+        results = results.drop(res_drop)
+        ci = ci.drop(ci_drop)
+    results = xr.concat([results, ci], 'results')
+    return results
+
+
+def bootstrap_hindcast(dple,
+                       lens,
+                       reference,
+                       metric='pearson_r',
+                       comparison='e2r',
+                       bootstrap=5,
+                       sig=95,
+                       pers_sig=None):
+    """See bootstrap_perfect_model but for hindcasts."""
+    if pers_sig is None:
+        pers_sig = sig
+
+    p = (100 - sig) / 100  # 0.05
+    ci_low = p / 2  # 0.025
+    ci_high = 1 - p / 2  # 0.975
+    p_pers = (100 - pers_sig) / 100  # 0.5
+    ci_low_pers = p_pers / 2
+    ci_high_pers = 1 - p_pers / 2
+    inits = dple.init.values
+    init = []
+    uninit = []
+    pers = []
+    # resample with replacement
+    # DoTo: parallelize loop
+    for _ in range(bootstrap):
+        smp = np.random.choice(inits, len(inits))
+        smp_dple = dple.sel(init=smp)
+        # compute init skill
+        init.append(
+            compute_reference(smp_dple,
```

Comment:
Places like this are the only time the function diverges. So maybe you assign a function to a variable. Like if it's `perfect_model`, then at the header you could do `skill_function=compute_reference` or `skill_function=compute_perfect_model`. Just an idea, but something to merge these two functions into a single set of code is important.

Reply:
adapted now.